ocaml/xapi/helpers.ml

```diff
@@ -291,6 +291,16 @@ let call_api_functions ~__context f =
          with e ->
            debug "Helpers.call_api_functions failed to logout: %s (ignoring)" (Printexc.to_string e))
 
+let test_mode = ref false
+let call_api_functions ~__context ?test_fn f =
```

Comment:
That's not how it's intended to work - I'll need to make this more obvious. The idea is to provide a function that will be executed in preference to `f` iff we're in test mode. So the callers of this function are distributed throughout the normal code of xapi, but calling API function doesn't work in the test context, so this is a mechanism to provide an alternative.

For example, in https://github.com/xapi-project/xen-api/pull/2970/files#diff-344dc1d17c4663add7fe5500813feef2R1622 - in the body of the event thread at some point we try to make an API call, which would fail in the test environment, whereas we can provide an alternative that _does_ work, but should only be used in testing. It's a gross hack, and we should be doing better, but that would be a _much_ bigger PR.